// @ngneat/transloco-locale/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of transloco-locale
 */
export { TranslocoLocaleService } from './lib/transloco-locale.service';
export { defaultConfig, LOCALE_DEFAULT_LOCALE, LOCALE_LANG_MAPPING, LOCALE_CONFIG, LOCALE_CURRENCY_MAPPING } from './lib/transloco-locale.config';
export {} from './lib/transloco-locale.types';
export { pipes, TranslocoLocaleModule } from './lib/transloco-locale.module';
export { TranslocoDatePipe } from './lib/pipes/transloco-date.pipe';
export { TranslocoCurrencyPipe } from './lib/pipes/transloco-currency.pipe';
export { TranslocoPercentPipe } from './lib/pipes/transloco-percent.pipe';
export { TranslocoDecimalPipe } from './lib/pipes/transloco-decimal.pipe';
export { TRANSLOCO_DATE_TRANSFORMER, TRANSLOCO_NUMBER_TRANSFORMER, DefaultDateTransformer, DefaultNumberTransformer } from './lib/transloco-locale.transformers';
